File: src/claude_gates/web_launcher.py
"""
web_launcher.py — SessionStart hook: idempotent launch of web dashboard server.

Probes /health before spawning. Writes PID file. Platform-aware detach.
Always fail-open (returns {}).
"""
from __future__ import annotations

import json
import os
import subprocess
import sys
from urllib.request import urlopen
from urllib.error import URLError

# ---------------------------------------------------------------------------
# Constants
# ---------------------------------------------------------------------------

_PLUGIN_ROOT: str = os.environ.get(
    "CLAUDE_PLUGIN_ROOT",
    os.path.dirname(os.path.dirname(os.path.dirname(os.path.abspath(__file__)))),
)

PORT: int = int(os.environ.get("CLAUDE_GATES_PORT", "64735"))

PID_FILE: str = os.path.join(os.getcwd(), ".sessions", ".webui.pid")

SCRIPT: str = os.path.join(_PLUGIN_ROOT, "scripts", "WebServer.py")


# ---------------------------------------------------------------------------
# Public API
# ---------------------------------------------------------------------------

def launch(data: dict) -> dict:
    """SessionStart handler — idempotently launch the web dashboard server."""
    try:
        if _health_check(PORT):
            return {}

        _clean_stale_pid()

        sessions_dir = os.path.dirname(PID_FILE)
        os.makedirs(sessions_dir, exist_ok=True)

        child = _spawn_server()

        if child.pid:
            with open(PID_FILE, "w") as f:
                f.write(str(child.pid))
    except Exception:
        pass  # fail-open

    return {}


def _health_check(port: int) -> bool:
    """HTTP GET http://127.0.0.1:{port}/health with 1-second timeout.

    Returns True only if response JSON has app == "claude-gates".
    """
    url = f"http://127.0.0.1:{port}/health"
    try:
        with urlopen(url, timeout=1) as resp:
            body = resp.read()
        data = json.loads(body)
        return data.get("app") == "claude-gates"
    except Exception:
        return False


def _clean_stale_pid() -> None:
    """Remove PID file if the recorded process is dead.

    Uses os.kill(pid, 0) to probe liveness.
    - OSError with errno == ESRCH (no such process) => dead, delete file.
    - PermissionError => process alive (Windows behavior), leave file.
    - No PID file => no-op.
    """
    try:
        with open(PID_FILE) as f:
            pid = int(f.read().strip())
    except (FileNotFoundError, ValueError):
        return

    try:
        os.kill(pid, 0)
        # Process is alive — health check should have caught this; leave the file.
    except PermissionError:
        # Windows raises PermissionError when process is alive but not owned.
        # Treat as alive.
        pass
    except OSError:
        # Process is dead — remove stale PID file.
        try:
            os.unlink(PID_FILE)
        except FileNotFoundError:
            pass


# ---------------------------------------------------------------------------
# Internal helpers
# ---------------------------------------------------------------------------

def _spawn_server() -> subprocess.Popen:
    """Spawn WebServer.py as a detached background process."""
    cmd = [sys.executable, SCRIPT]
    env = {**os.environ, "CLAUDE_GATES_PORT": str(PORT)}

    if sys.platform == "win32":
        return subprocess.Popen(
            cmd,
            cwd=os.getcwd(),
            env=env,
            stdin=subprocess.DEVNULL,
            stdout=subprocess.DEVNULL,
            stderr=subprocess.DEVNULL,
            creationflags=(
                subprocess.CREATE_NEW_PROCESS_GROUP | subprocess.DETACHED_PROCESS
            ),
        )
    else:
        return subprocess.Popen(
            cmd,
            cwd=os.getcwd(),
            env=env,
            stdin=subprocess.DEVNULL,
            stdout=subprocess.DEVNULL,
            stderr=subprocess.DEVNULL,
            start_new_session=True,
        )

File: src/WebLauncher.ts
#!/usr/bin/env node
/**
 * Pipeline Web UI — SessionStart launcher.
 *
 * Idempotent: probes /health before spawning. Writes PID file.
 * Always exits 0 (fail-open).
 */

import { spawn, } from "child_process";
import fs from "fs";
import http from "http";
import path from "path";

const SESSIONS_DIR = path.join(process.cwd(), ".sessions",);
const PID_FILE = path.join(SESSIONS_DIR, ".webui.pid",);
const PORT = parseInt(process.env.CLAUDE_GATES_PORT || "64735", 10,);
const SCRIPT = path.join(__dirname, "WebServer.js",);

function healthCheck(): Promise<boolean>
{
  return new Promise((resolve,) =>
  {
    const req = http.get(`http://127.0.0.1:${PORT}/health`, (res,) =>
    {
      let body = "";
      res.on("data", (chunk: Buffer,) =>
      {
        body += chunk;
      },);
      res.on("end", () =>
      {
        try
        {
          const data = JSON.parse(body,);
          resolve(data.app === "claude-gates",);
        }
        catch
        {
          resolve(false,);
        }
      },);
    },);
    req.on("error", () => resolve(false,),);
    req.setTimeout(1000, () =>
    {
      req.destroy();
      resolve(false,);
    },);
  },);
}

function cleanStalePid(): void
{
  try
  {
    const pid = parseInt(fs.readFileSync(PID_FILE, "utf-8",).trim(), 10,);
    if (pid)
    {
      try
      {
        process.kill(pid, 0,); // test if alive
        return; // process is alive, leave it
      }
      catch
      {
        // process is dead, remove stale PID file
      }
    }
    fs.unlinkSync(PID_FILE,);
  }
  catch
  {
    // no PID file
  }
}

async function main(): Promise<void>
{
  // Already running?
  const alive = await healthCheck();
  if (alive)
  {
    process.exit(0,);
  }

  // Clean stale PID
  cleanStalePid();

  // Ensure sessions dir exists
  fs.mkdirSync(SESSIONS_DIR, { recursive: true, },);

  // Spawn detached server
  const child = spawn(process.execPath, [SCRIPT,], {
    cwd: process.cwd(),
    detached: true,
    stdio: "ignore",
    env: { ...process.env, CLAUDE_GATES_PORT: String(PORT,), },
  },);

  child.unref();

  if (child.pid)
  {
    fs.writeFileSync(PID_FILE, String(child.pid,), "utf-8",);
  }

  process.exit(0,);
}

main().catch(() => process.exit(0,));
